sympy/functions/elementary/piecewise.py

```diff
@@ -1238,11 +1238,8 @@ def piecewise_simplify(expr, **kwargs):
             if eqs and not other:
                 eqs = list(ordered(eqs))
                 for e in eqs:
-                    # these blessed lhs objects behave like Symbols
-                    # and the rhs are simple replacements for the "symbols"
-                    if _blessed(e):
```

Comment:
There must have been a reason for adding this check. Some substitutions aren't valid. I thought maybe that substituting for an ExprCondPair in a Piecewise would fail. Has that been changed?

Reply:
> I thought maybe that substituting for an ExprCondPair in a Piecewise would fail. Has that been changed?
```
eq = Piecewise((x, Eq(y, 0)), (y, True))
print(eq.subs([(x, 2), (y, 1)]))
```
This does work.
> There must have been a reason for adding this check.

I understand this but the `_blessed()` condition checks on the basis of `_diff_wrt` which is False by default for all expressions.

Reply:
Can you give some examples which should not be valid ?

Reply:
There are lots of possible invalid substitutions e.g.:
```julia
In [1]: Piecewise((x, y<1), (z, True))                                                                                                         
Out[1]: 
⎧x  for y < 1
⎨            
⎩z  otherwise

In [2]: Piecewise((x, y<1), (z, True)).subs(y < 1, exp(2))                                                                                     
---------------------------------------------------------------------------
TypeError
```
Here we are substituting an `Expr` in place of a `Boolean` and `Piecewise` rightly complains.

I would investigate this by using git blame to find the commit/PR where this piece of code was introduced and who the author was or who merged it. The discussion in the PR might explain the reason. Also you could CC the author giving them a chance to explain.

Reply:
The example that you showed does raise a TypeError for me as well. Also I found the commit and I get https://github.com/sympy/sympy/commit/9d3fa8b4d4963c059c139f3860b8cc1d611adbae
It appears there is something wrong with the commit/blame history as mentioned in the PR
https://github.com/sympy/sympy/pull/18488 and https://github.com/sympy/sympy/issues/19181.

Reply:
It looks like that commit has ruined the history in this file

Reply:
You can find the blame from before that commit here:
https://github.com/sympy/sympy/blame/d5df2bd6a274db8b8d3a5cf2622d16dfb3f3d8a4/sympy/functions/elementary/piecewise.py

Reply:
For sake of simplicity it is attempting only straightforward subs that are in the form (s1, s2) or (s, ns) or (s, r) where s is a symbol-like entity and ns and r are number symbol and rational, respectively. It was added in #17259

Reply:
As this does not work on instances of `Expr`, would removing the condition be fine?

Reply:
@smichr It seems that in the PR you mentioned `piecewise_simplify()` was introduced and did the same work as `_eval_simplify()`. I think the condition specifically was modified/added in https://github.com/sympy/sympy/commit/5e34c1b1370d5996c10cb2e72841cbc752cb419f

Reply:
substitution of non-symbol-like entities is not as simple as for the others. What test fails if you leave this optimization in place? Is there an impact on speed?

Reply:
None of the tests fail. I used timeit for measuring the execution time and ran it multiple times. The results I got are: 
With `_blessed()`: 0.3315324999999998 , 0.3113937
without : 0.35980409999999985 , 0.32984559999999874
Although the condition helps with speed I am not sure if it is significant enough.

Reply:
@smichr @oscarbenjamin Is it merge able now?

Reply:
I would like to see the use of `_blessed` retained. It is used above and it is used here, likely, for the same reason. So without good reason to do otherwise I would like to keep it.

Reply:
The condition prevents from any non-symbol like objects (like the ones mentioned in the issue) to be simplified. If removing it isn't recommended, I think I could be modified to include `Expr` objects as well. Do you think such a change would be fine?

Reply:
> like the ones mentioned in the issue

One of these expressions should be added as a test because nothing currently fails if this code is reverted back to using `_blessed`.

Reply:
On the local branch same as #19487 where `_eval_rewrite_as_Abs()` is implemented and `_blessed` is used, the following are the results:
`Abs(x) * sign(x)`  gives `|𝑥|sign(𝑥)`  and

`simplify(Piecewise((x*Abs(x + f(x)**3), Eq(x + f(x)**3, 0)),(x*Abs(x + f(x)**3) + x + f(x)**3, True)))` gives
`Piecewise((x*Abs(x + f(x)**3), Eq(x + f(x)**3, 0)), (x*Abs(x + f(x)**3) + x + f(x)**3, True))`

On this branch these cases are simplified.

Reply:
I added the test that would fail if the `_blessed` was not removed. I wonder if a compromise would be to only do this `if len(args) == 2 or _blessed(e):`

Reply:
I don't think `len(args) ==2` would work for a general case. I tried 
```
_blessed = lambda e: (getattr(e.lhs, '_diff_wrt', False) or isinstance(e.lhs, Expr)) and (
        getattr(e.rhs, '_diff_wrt', None) or
        isinstance(e.rhs, (Rational, NumberSymbol, Expr)))
```
Although it works but I think it is very similar to removing the condition.

Reply:
I wasn't thinking to modify `_blessed`, just to make the change roosed below.

Reply:
I applied the suggestion...let's see if that passes.

Reply:
I think it would pass but `len(args) == 2` would work only for Piecewise containing 2 arguments. What about a general case for any number of arguments?